src/Week2/LogAnalysis.hs

```diff
@@ -0,0 +1,161 @@
+ can use the read function. Other functions which may (or may not)
+ be useful to you include lines, words, unwords, take, drop, and (.).
+-}
+parseMessage :: String -> LogMessage
+parseMessage msg = case words msg of
+  ("I" : timestamp : info) -> case (readMaybe timestamp :: Maybe Int) > Just 0 of
+    True -> LogMessage Info (read timestamp) (unwords info)
+    _ -> Unknown msg
+  ("W" : timestamp : info) -> case (readMaybe timestamp :: Maybe Int) > Just 0 of
+    True -> LogMessage Warning (read timestamp) (unwords info)
+    _ -> Unknown msg
+  ("E" : level : timestamp : info) -> case (readMaybe level :: Maybe Int) > Just 0
+                                      && (readMaybe timestamp :: Maybe Int) > Just 0 of
+    True -> LogMessage (Error $ read level) (read timestamp) (unwords info)
+    _ -> Unknown msg
+  info -> Unknown (unwords info)
+
+
+-- Parse an entire log file at once and return its contents as a list of LogMessages
+parse :: String -> [LogMessage]
+parse = map (parseMessage) . lines
+
+
+-------- Putting the logs in order ------------
+
+{-
+ Exercise 2: Define a function
+    insert :: LogMessage -> MessageTree -> MessageTree
+
+ which inserts a new LogMessage into an existing MessageTree, producing a new MessageTree.
+ Insert may assume that it is given a sorted MessageTree, and must produce a new sorted
+ MessageTree containing the new LogMessage in addition to the contents of the
+ original MessageTree.
+
+ However, note that if insert is given a LogMessage which is Unknown, it should
+ return the MessageTree unchanged.
+-}
+insert :: LogMessage -> MessageTree -> MessageTree
+insert x Leaf = Node Leaf x Leaf
+insert (Unknown _) tree = tree
+insert _ tree@(Node _ (Unknown _) _) = tree
+insert msg1@(LogMessage _ t1 _ ) tree@(Node left msg2@(LogMessage _ t2 _) right)
+  | t2 > t1   = Node (insert msg1 left) msg2 right
+  | t2 < t1   = Node left msg2 (insert msg1 right)
+  | otherwise = tree
+
+
+{-
+ Exercise 3: Once we can insert a single LogMessage into a MessageTree,
+ we can build a complete MessageTree from a list of messages. Specifically, define a function
+    build :: [LogMessage] -> MessageTree
+
+ which builds up a MessageTree containing the messages in the list,
+ by successively
+-}
+build :: [LogMessage] -> MessageTree
+build = foldr insert Leaf
+
+
+{-
+ Exercise 4: Finally, define the function
+   inOrder :: MessageTree -> [LogMessage]
+
+ which takes a sorted MessageTree and produces a list of all the
+ LogMessages it contains, sorted by timestamp from smallest to biggest.
+ (This is known as an in-order traversal of the MessageTree.)
+
+ With these functions, we can now remove Unknown messages and
+ sort the well-formed messages using an expression such as:
+   inOrder (build tree)
+-}
+inOrder :: MessageTree -> [LogMessage]
+-- base case
+inOrder Leaf                  = []
+-- Visit in the order: Left subtree -> root -> right subtree
+inOrder (Node left msg right) = inOrder left ++ [msg] ++ inOrder right
+
```

Comment:
This is the correct solution. However, we still have the issue of expensive list concatenation operations. Can you think of a more elegant solution that _does not_ use the `++` operator? (it will involve a helper function)

Reply:
I've added an alternate implementation to this.
I've tried to have a helper function in the `where` clause of the function itself. Can you check and let me know if that's what you expected?